src/platform/i386/chal_pgtbl.c

```diff
@@ -0,0 +1,514 @@
+	if (!pgtbl_ispresent(*flags)) return NULL;
+	return ret;
+}
+
+unsigned long *
+chal_pgtbl_lkup_pte(pgtbl_t pt, u32_t addr, u32_t *flags)
+{
+	return __pgtbl_lkupan((pgtbl_t)((unsigned long)pt | X86_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT, PGTBL_DEPTH,
+	                      flags);
+}
+
+int
+chal_pgtbl_get_cosframe(pgtbl_t pt, vaddr_t frame_addr, paddr_t *cosframe)
+{
+	u32_t          flags;
+	unsigned long *pte;
+	paddr_t        v;
+
+	pte = pgtbl_lkup_pte(pt, frame_addr, &flags);
+	if (!pte) return -EINVAL;
+
+	v = *pte;
+	if (!(v & X86_COSFRAME)) return -EINVAL;
+
+	*cosframe = v & PGTBL_FRAME_MASK;
+
+	return 0;
+}
+
+pgtbl_t
+chal_pgtbl_create(void *page, void *curr_pgtbl)
+{
+	pgtbl_t ret = pgtbl_alloc(page);
+	/* Copying the kernel part of the pgd. */
+	memcpy(page + KERNEL_PGD_REGION_OFFSET, (void *)chal_pa2va((paddr_t)curr_pgtbl) + KERNEL_PGD_REGION_OFFSET,
+	       KERNEL_PGD_REGION_SIZE);
+
+	return ret;
+}
+
+int
+chal_pgtbl_quie_check(u32_t orig_v)
+{
+	livenessid_t lid;
+	u64_t        ts;
+
+	if (orig_v & X86_QUIESCENCE) {
+		lid = orig_v >> PGTBL_PAGEIDX_SHIFT;
+		/* An unmap happened at this vaddr before. We need to
+		 * make sure that all cores have done tlb flush before
+		 * creating new mapping. */
+		assert(lid < LTBL_ENTS);
+
+		if (ltbl_get_timestamp(lid, &ts)) return -EFAULT;
+		if (!tlb_quiescence_check(ts)) {
+			printk("kern tsc %llu, lid %d, last flush %llu\n", ts, lid,
+			       tlb_quiescence[get_cpuid()].last_periodic_flush);
+			return -EQUIESCENCE;
+		}
+	}
+
+	return 0;
+}
+
+void
+chal_pgtbl_init_pte(void *pte)
+{
+	int            i;
+	unsigned long *vals = pte;
+
+	for (i = 0; i < (1 << PGTBL_ORD); i++) vals[i] = 0;
+}
+
+int
+chal_pgtbl_cons(struct cap_captbl *ct, struct cap_captbl *ctsub, capid_t expandid, unsigned long depth)
+{
+	/* FIXME: we need to ensure TLB quiescence for pgtbl cons/decons! */
+	u32_t flags = 0, old_pte, new_pte, old_v, refcnt_flags;
+	unsigned long *    intern;
+	int                ret = 0;
+
+	intern = pgtbl_lkup_lvl(((struct cap_pgtbl *)ct)->pgtbl, expandid, &flags, ct->lvl, depth);
+	if (!intern) return -ENOENT;
+	old_pte = *intern;
+	if (pgtbl_ispresent(old_pte)) return -EPERM;
+
+	old_v = refcnt_flags = ((struct cap_pgtbl *)ctsub)->refcnt_flags;
+	if (refcnt_flags & CAP_MEM_FROZEN_FLAG) return -EINVAL;
+	if ((refcnt_flags & CAP_REFCNT_MAX) == CAP_REFCNT_MAX) return -EOVERFLOW;
+
+	refcnt_flags++;
+	ret = cos_cas((unsigned long *)&(((struct cap_pgtbl *)ctsub)->refcnt_flags), old_v, refcnt_flags);
+	if (ret != CAS_SUCCESS) return -ECASFAIL;
+
+	new_pte = (u32_t)chal_va2pa(
+	          (void *)((unsigned long)(((struct cap_pgtbl *)ctsub)->pgtbl) & PGTBL_FRAME_MASK))
+	          | X86_INTERN_DEF;
+
+	ret = cos_cas(intern, old_pte, new_pte);
+	if (ret != CAS_SUCCESS) {
+		/* decrement to restore the refcnt on failure. */
+		cos_faa((int *)&(((struct cap_pgtbl *)ctsub)->refcnt_flags), -1);
+		return -ECASFAIL;
+	} else {
+		ret = 0;
+	}
+
+	return ret;
+}
+
```

Comment:
I trust that for this file you did a `git mv` before you updated it so that git tracks its history?